train.py

```diff
@@ -0,0 +1,356 @@
+import os
+import random
+
+import hydra
+import torch
+import wandb
+import torch.nn as nn
+from tqdm import tqdm
+import torch.optim as optim
+import torch.nn.functional as F
+from omegaconf import OmegaConf
+from huggingface_hub import upload_file
+from omegaconf.listconfig import ListConfig
+from torch.utils.data import Subset, DataLoader
+from datasets import load_dataset, concatenate_datasets
+from huggingface_hub.file_download import hf_hub_download
+
+from data.dataset import MidiDataset
+from models.reverse_diffusion import Unet
+from models.ema import ExponentialMovingAverage
+from models.forward_diffusion import ForwardDiffusion
+from models.velocity_time_encoder import VelocityTimeEncoder
+
+
+def makedir_if_not_exists(dir: str):
+    if not os.path.exists(dir):
+        os.makedirs(dir)
+
+
+def preprocess_dataset(dataset_name: str | list[str], batch_size: int, num_workers: int, *, overfit_single_batch: bool = False):
+    hf_token = os.environ["HUGGINGFACE_TOKEN"]
+
+    if isinstance(dataset_name, (list, ListConfig)):
+        train_ds = []
+        val_ds = []
+        test_ds = []
+
+        for ds_name in dataset_name:
+            tr_ds = load_dataset(ds_name, split="train", use_auth_token=hf_token)
+            v_ds = load_dataset(ds_name, split="validation", use_auth_token=hf_token)
+            t_ds = load_dataset(ds_name, split="test", use_auth_token=hf_token)
+
+            train_ds.append(tr_ds)
+            val_ds.append(v_ds)
+            test_ds.append(t_ds)
+
+        train_ds = concatenate_datasets(train_ds)
+        val_ds = concatenate_datasets(val_ds)
+        test_ds = concatenate_datasets(test_ds)
+    else:
+        train_ds = load_dataset(dataset_name, split="train", use_auth_token=hf_token)
+        val_ds = load_dataset(dataset_name, split="validation", use_auth_token=hf_token)
+        test_ds = load_dataset(dataset_name, split="test", use_auth_token=hf_token)
+
+    train_ds = MidiDataset(train_ds, augmentation_percentage=0.8)
+    val_ds = MidiDataset(val_ds, augmentation_percentage=0.0)
+    test_ds = MidiDataset(test_ds, augmentation_percentage=0.0)
+
+    if overfit_single_batch:
+        train_ds = Subset(train_ds, indices=range(batch_size))
+        val_ds = Subset(val_ds, indices=range(batch_size))
+        test_ds = Subset(test_ds, indices=range(batch_size))
+
+    # dataloaders
+    train_dataloader = DataLoader(train_ds, batch_size=batch_size, num_workers=num_workers, shuffle=True)
+    val_dataloader = DataLoader(val_ds, batch_size=batch_size, num_workers=num_workers)
+    test_dataloader = DataLoader(test_ds, batch_size=batch_size, num_workers=num_workers)
+
+    return train_dataloader, val_dataloader, test_dataloader
+
+
+def forward_step(
+    model: Unet,
+    forward_diffusion: ForwardDiffusion,
+    conditioning_model: nn.Module,
+    batch: dict[str, torch.Tensor, torch.Tensor],
+    device: torch.device,
+) -> float:
+    x = batch["velocity"].to(device)
```

Comment:
```suggestion
    velocity = batch["velocity"].to(device)
```

It's almost always better to have meaningful names :)